### PR TITLE
Move bad coordinates check

### PR DESCRIPTION
**Pull request by ronlevine**
### Description

Follow up to #908.
Move the bad coordinates check to the root method, `public Iterator query(final String reg)`. If this public method is used by an application, the bad coordinates will return an `EOF_ITERATOR;`

### Checklist

- [X] Code compiles correctly
- [X] New tests covering changes and new functionality
- [X] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

**Comment by ronlevine** [comment]
@yfarjoun Please review.

File: src/test/java/htsjdk/tribble/readers/TabixReaderTest.java [hunk 1]
```diff
@@ -95,7 +98,15 @@ public void testIterators() throws IOException {
         iter=tabixReader.query("1",Integer.MAX_VALUE-1,Integer.MAX_VALUE);
         Assert.assertNotNull(iter);
         Assert.assertNull(iter.next());
-        
+
+        iter=tabixReader.query("1",-1,Integer.MAX_VALUE);
```
Comment (yfarjoun) [inline src/test/java/htsjdk/tribble/readers/TabixReaderTest.java]: spaces around operators

**Comment by yfarjoun** [comment]
@ronlevine are you still available to process this? 

if not @lbergelson what's the procedure?

**Comment by yfarjoun** [comment]
@cmnbroad can you take over this?

**Comment by cmnbroad** [comment]
@yfarjoun Sure.

**Comment by cmnbroad** [comment]
@yfarjoun I don't see anything in this PR that addresses the second issue raised in the ticket. A couple of questions:

Do you think we should we retain the current behavior where querying a non-existent sequence name silently returns an empty iterator ? It seems to me it should throw. Same for start index <= 0. Such queries seem incoherent, but the caller has no way to notify the user. It would be a behavior change though.
It looks like the parseReg method is converting com 1-based to 0-based, and subtracts 1 from the start, but not the end. Unless its intentionally converting from a fully closed interval to half-closed, half-open, that seems wrong. Do you know if thats intentional ? I'd like to add some comments/javadoc and boundary condition tests, but I'm not sure what the intentions are.

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 2]
```diff
@@ -439,7 +439,7 @@ public String next() throws IOException {
     public Iterator query(final int tid, final int beg, final int end) {
         TPair64[] off, chunks;
         long min_off;
-        if(tid< 0 || tid>=this.mIndex.length) return EOF_ITERATOR;
+        if (tid < 0 || beg < 0 || end < 0 || tid >= this.mIndex.length) return EOF_ITERATOR;
```
Comment (droazen) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: I think this should be `end <= 0`, since the coordinates are 0-based closed-open.

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 3]
```diff
@@ -430,7 +430,7 @@ public String next() throws IOException {
     }
 
     /**
-     * Return
+     * Get an iterator for an interval specified by the sequence id and begin and end coordinates
      * @param tid Sequence id
      * @param beg beginning of interval, genomic coords
      * @param end end of interval, genomic coords
```
Comment (droazen) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: The javadoc for this method should document the convention being used for the genomic coordinates. Based on the callsites for this method, they appear to be BED-style (0-based, closed-open).

File: src/test/java/htsjdk/tribble/readers/TabixReaderTest.java [hunk 4]
```diff
@@ -95,7 +98,15 @@ public void testIterators() throws IOException {
...
+        Assert.assertNotNull(iter);
+        Assert.assertNull(iter.next());
+
+        iter=tabixReader.query("1",Integer.MAX_VALUE,-1);
+        Assert.assertNotNull(iter);
+        Assert.assertNull(iter.next());
+
```
Comment (droazen) [inline src/test/java/htsjdk/tribble/readers/TabixReaderTest.java]: Can we add a test case in which end == 0?

Comment (droazen) [reply src/test/java/htsjdk/tribble/readers/TabixReaderTest.java]: Also add test cases for the other overloads of `query()`, since they have difference interval conventions (the String version is 1-based closed, the 3-arg version is 0-based closed-open!)

File: src/main/java/htsjdk/tribble/readers/TabixReader.java (hunk 3 above)
Comment (droazen) [reply src/main/java/htsjdk/tribble/readers/TabixReader.java]: Also applies to the other overloads of `query()`. It looks like some overloads use the 1-based closed convention, while others like this one use the 0-based closed-open convention.

**Comment by droazen** [comment]
Review complete, back to @cmnbroad

File: src/main/java/htsjdk/tribble/readers/TabixReader.java (hunk 3 above)
Comment (droazen) [reply src/main/java/htsjdk/tribble/readers/TabixReader.java]: Also document the behavior with non-existent contigs, etc.

**Comment by yfarjoun** [comment]
ping....

**Comment by fleharty** [comment]
@yfarjoun I'm taking this over, but don't have the ability to assign since I don't have commit privileges.

**Comment by yfarjoun** [comment]
@droazen ? @cmnbroad ? who's court is this in now?

**Comment by fleharty** [comment]
@yfarjoun my court

**Comment by yfarjoun** [comment]
looks good now.
